File: python/31_NextPermutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        length = len(nums)
        for i in range(length-2, -1, -1):
            if nums[i] < nums[i+1]:
                for j in range(i+1, length):
                    if nums[j] > nums[i] and j == length-1 or nums[j] >= nums[i] >= nums[j+1]:
                        nums[i], nums[j] = nums[j], nums[i]
                        nums[i+1:] = nums[:i:-1]
                        return nums
        nums[:] = nums[::-1]
        return nums

File: cpp/31_NextPermutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int N = nums.size();
        int start = -1, end = N - 1;
        for (int i = N - 2; i >= 0; --i) {
            if (nums[i] < nums[i+1]) {
                int small = INT_MAX;
                start = i;
                for (int j = i + 1; j < N; ++j) {
                    if (nums[i] < nums[j] && nums[j] <= small) {
                        small = nums[j];
                        end = j;
                    }
                }
                break;
            }
        }
        
        if (start >= 0)
            swap(nums, start, end);
        
        reverse(nums, start + 1, N - 1);
    }
    
    void swap(vector<int>& nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
    
    void reverse(vector<int>& nums, int i, int j) {
        while (i <= j) {
            swap(nums, i, j);
            i++, j--;
        }
    }
};
